design-docs/commands-v3.md

```diff
@@ -0,0 +1,305 @@
+*own* the elevator and coral resources for the entire sequence, but not control the elevator during
+the coral scoring section, nor control the coral scoring mechanism while the elevator is moving.
+
+The v2 framework worked around this problem with so-called proxy commands, which have no
+requirements of their own and simply schedule and await the *real* command. This removed the
+requirement for their used subsystems from the parent group, and allowed for the subsystem’s default
+command to run after the proxied command completed. It also meant that *every* command that required
+that subsystem would also need to be proxied; otherwise, the group would still have ownership of the
+subsystem, and the proxied command would interrupt it (and thus itself).
+
+The v3 framework allows nested commands to use resources not required by the parent command;
+however, the built-in parallel and sequential groups will take full ownership of all resources
+required by their nested commands for safety and to keep behavior parity with previous command
+frameworks.
+
+```java
+public Command outerCommand() {
+  // The outer command only requires the outer resource
+  return run(coroutine -> {
+    // But can schedule a command that requires any other resource
+    coroutine.await(innerResource.innerCommand());
+    // And only requires that inner resource while its command runs
+    coroutine.await(otherResource().otherCommand());
+  }).named("Outer");
+}
+```
+
+Scheduling a command that requires an inner resource will also cancel its parent, even if the parent
+does not require that inner resource. Using the above example, directly scheduling a command that
+requires `innerResource` will cancel a running `outerCommand` - but only if `outerCommand` is
+*currently using* the inner resource at the time.
+
+### Priority Levels
+
+Priority levels allow for finer-grained control over when commands should be interruptible than a
+binary interrupt/ignore interrupt flag. This can be particularly helpful for LED control, where
+teams use lights to indicate robot activity or error states with error indicators taking priority.
+
+### Suspend/Resume
+
+Allow commands to be temporarily paused while a higher priority or interrupting command runs, then
+be automatically resumed after that higher priority command finishes. Note that resuming commands
+must check for a condition to determine if they should still be running, since the time between
+pause and resume can be arbitrarily long.
+
+Suspending a command will suspend all its children, regardless of if those children have declared
+themselves suspendable or not. Likewise, cancelling a non-suspendable command will also cancel all
+its children, even if some of those children are suspendable.
+
+Suspending a command that already has suspended children will still suspend everything; however,
+upon resume, those suspended children will remain suspended.
```

Comment:
This is going to be very difficult to implement. Probably won't make it into the initial version of the framework (this PR), but can be done in a followup with a limited scope